cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/index.tsx

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import IconButton from '@material-ui/core/IconButton';
+import { withStyles } from '@material-ui/core/styles';
+import AddIcon from '@material-ui/icons/Add';
+import HierarchyTree from './HierarchyTree';
+import uuidV4 from 'uuid/v4';
+import { objectQuery } from 'services/helpers';
+import { flatToTree, toJsonHandler, jsonToFlat, getFields, inputSchemaToFlat } from './dataHandler';
+
+export const IconWrapper = withStyles(() => {
+  return {
+    root: {
+      borderRadius: 10,
+      padding: '5px',
+      marginTop: '5px',
+      '&:focus': {
+        outline: 'none',
+      },
+    },
+  };
+})(IconButton);
+
+const HierarchyWidget = (props) => {
+  const [records, setRecords] = React.useState([]);
+  const [tree, setTree] = React.useState([]);
+  const [jsonData, setJsonData] = React.useState({});
+  const inputSchema = objectQuery(props.extraConfig, 'inputSchema') || [];
+  const allowedTypes: string[] = props.widgetProps.allowedTypes || [];
+  const fieldValues = getFields(inputSchema, allowedTypes);
+  const dropdownData = inputSchemaToFlat(fieldValues);
+
+  React.useEffect(() => {
+    if (props.value) {
+      if (JSON.parse(props.value).length !== 0) {
+        const jsonToFlatData = jsonToFlat(JSON.parse(props.value));
+        jsonToFlatData.map((record) => {
+          if (record.path) {
+            const el = dropdownData.find((item) => item.name === record.name);
+            record.type = el.type;
+          }
+        });
+        setRecords(jsonToFlatData);
+      }
+    }
+  }, []);
+
+  React.useEffect(() => {
+    setTree(flatToTree(typeof records === 'string' ? JSON.parse(records) : records));
+    props.onChange(JSON.stringify(jsonData));
+  }, [records]);
```

Comment:
This will be a little bit costly comparison/operation. Have an `onChange` method to `HierarchyTree` and have it call the callback when the value changes.

This should follow the default widget pattern where each widget will have a value and an onChange handler when the value changes.